src/stream/src/executor/backfill.rs

```diff
@@ -0,0 +1,380 @@
+///
+/// It can only buffer chunks between two (checkpoint) barriers instead of unbundled memory usage of
+/// `RearrangedChainExecutor`.
+///
+/// It uses the latest epoch to read the snapshot of the upstream mv during two barriers and all the
+/// `StreamChunk` of the snapshot read will forward to the downstream.
+///
+/// It uses `current_pos` to record the progress of the backfill (the pk of the upstream mv) and
+/// `current_pos` is initiated as an empty `Row`.
+///
+/// All upstream messages during the two barriers interval will be buffered and decide to forward or
+/// ignore based on the `current_pos` at the end of the later barrier. Once `current_pos` reaches
+/// the end of the upstream mv pk, the backfill would finish.
+///
+/// Notice: The pk we talk about here refers to the storage primary key.
+pub struct BackfillExecutor<S: StateStore> {
+    /// Upstream table
+    table: StorageTable<S>,
+    /// Upstream with the same schema with the upstream table.
+    upstream: BoxedExecutor,
+
+    /// The column indices need to be forwarded to the downstream.
+    upstream_indices: Arc<[usize]>,
+
+    /// Current position of the table storage primary key.
+    /// None means it starts from the beginning.
+    current_pos: Option<Row>,
+
+    progress: CreateMviewProgress,
+
+    actor_id: ActorId,
+
+    info: ExecutorInfo,
+}
+
+impl<S> BackfillExecutor<S>
+where
+    S: StateStore,
+{
+    pub fn new(
+        table: StorageTable<S>,
+        upstream: BoxedExecutor,
+        upstream_indices: Vec<usize>,
+        progress: CreateMviewProgress,
+        schema: Schema,
+        pk_indices: PkIndices,
+    ) -> Self {
+        Self {
+            info: ExecutorInfo {
+                schema,
+                pk_indices,
+                identity: "BackfillExecutor".to_owned(),
+            },
+            table,
+            upstream,
+            upstream_indices: upstream_indices.into(),
+            current_pos: None,
+            actor_id: progress.actor_id(),
+            progress,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn execute_inner(mut self) {
+        // Table storage primary key.
+        let table_pk_indices = self.table.pk_indices().to_vec();
+        let upstream_indices = self.upstream_indices.to_vec();
+
+        let mut upstream = self.upstream.execute();
+
+        // Poll the upstream to get the first barrier.
+        let first_barrier = expect_first_barrier(&mut upstream).await?;
+        let init_epoch = first_barrier.epoch.prev;
+
+        // If the barrier is a conf change of creating this mview, init backfill from its epoch.
+        // Otherwise, it means we've recovered and the backfill is already finished.
+        let to_backfill = first_barrier.is_add_dispatcher(self.actor_id);
+
+        // The first barrier message should be propagated.
+        yield Message::Barrier(first_barrier.clone());
+
+        if to_backfill {
```

Comment:
We may return early if not backfill to make the indent lighter. (Also appliable to rearranged chain, but I don't want to introduce this whitespace-only diff. 🤣)
```suggestion
        if !to_backfill {
            ...
            return;
        }
```

Reply:
Yes, I have tried before, but it seems the compiler isn't clear enough to detect it it ownership safe.